sympy/tensor/array/arrayop.py

```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
 
     """
     from sympy.matrices import MatrixBase
+    from sympy.tensor import SparseNDimArray
     array_types = (Iterable, MatrixBase, NDimArray)
-
     if isinstance(dx, array_types):
         dx = ImmutableDenseNDimArray(dx)
         for i in dx:
             if not i._diff_wrt:
                 raise ValueError("cannot derive by this array")
 
     if isinstance(expr, array_types):
-        expr = ImmutableDenseNDimArray(expr)
+        if isinstance(expr, NDimArray):
+            expr = expr.as_immutable()
+        else:
+            expr = ImmutableDenseNDimArray(expr)
+
         if isinstance(dx, array_types):
-            new_array = [[y.diff(x) for y in expr] for x in dx]
+            if isinstance(expr, SparseNDimArray):
+                new_array = _sparse_array_derive_by_array(expr, dx)
+            else:
+                new_array = [[y.diff(x) for y in expr] for x in dx]
             return type(expr)(new_array, dx.shape + expr.shape)
         else:
-            return expr.diff(dx)
+            if isinstance(expr, SparseNDimArray):
+                new_array = {k: v.diff(dx) for (k, v) in expr._sparse_array.items() if v.diff(dx)!= 0}
+                return type(expr)(new_array, expr.shape)
+            else:
+                return expr.diff(dx)
     else:
         if isinstance(dx, array_types):
             return ImmutableDenseNDimArray([expr.diff(i) for i in dx], dx.shape)
         else:
             return diff(expr, dx)
 
+def _sparse_array_derive_by_array(expr, dx):
+    # Perfom a derivative on a sparse array by each term of dx. Regroup the returned
+    # diff sprase array by each term to a dictionary.
+
+    sparse_array = {}
+    loop_size = len(expr)
+    for i, x in enumerate(dx):
+        sparse_diff = derive_by_array(expr, x)._sparse_array
```

Comment:
why is this calling derive_by_array?

Reply:
This is because `.diff()` was slow when it comes to large scale sparse array, so I chose not to use it. Besides, `derive_by_array `, as said in the description of the function, supports the derivate by scalars, so I used it. If `.diff()` or `applyfunc` is fixed, then we can simply call `expr.diff(x)`.